sdk/identity/identity/src/credentials/managedIdentityCredential/cloudShellMsi.ts

```diff
@@ -53,14 +53,21 @@ function prepareRequestOptions(
 
 /**
  * Defines how to determine whether the Azure Cloud Shell MSI is available, and also how to retrieve a token from the Azure Cloud Shell MSI.
+ * Since Azure Managed Identities aren't available in the Azure Cloud Shell, we log a warning for users that try to access cloud shell using user assigned identity.
  */
 export const cloudShellMsi: MSI = {
-  async isAvailable(scopes): Promise<boolean> {
+  async isAvailable(scopes, _identityClient, clientId): Promise<boolean> {
     const resource = mapScopesToResource(scopes);
     if (!resource) {
       logger.info(`${msiName}: Unavailable. Multiple scopes are not supported.`);
       return false;
     }
+    if (clientId) {
+      logger.warning(
+        `${msiName}: Unavailable. Azure Managed Identities aren't available in the Azure Cloud Shell.`
+      );
+      return false;
+    }
```

Comment:
I'm a bit confused by this. I think my confusion comes from three intersecting concerns:

1. The warning is logged in the `isAvailable` test. Doesn't that mean this warning will pretty much _always_ trigger if the user specifies a `managedIdentityClientId` when the ManagedIdentityCredential tries to cache its available MSI providers? I guess what I'm saying is that it feels weird to have a warning-level log that will happen any time you provide a client ID.

2. The warning says "managed identities aren't available" in the cloud shell, but really the problem is that you can't _pick_ the assigned managed identity, as the issue title says, is that right? I think the warning message could mislead someone into thinking that no form of Managed Identity is supported in the cloud shell.

3. Previously, we just ignored this option, correct? So, what if the user had passed in the client ID of the identity that their cloud shell instance is actually assigned (is that possible)? We'd log a warning and return `false` now, even though the option was semantically correct, but ignored and _appeared_ to work previously. Feels a bit like a breaking change.

CC @sadasant WDYT?

Reply:
> Doesn't that mean this warning will pretty much always trigger if the user specifies a managedIdentityClientId

Yep. However, SDK logs are not automatically logged. They’re only logged when they’re enabled.

> warning-level log that will happen any time you provide a client ID

We have a bunch of info level logs which are far more pervasive. When I think of whether this should be an info level log or a warning level log, I think this is most likely a warning level log, since if the user provided a custom client ID *and* they’re in the cloud shell environment, an info log may not be sufficient indication that there was something wrong with the authentication.

The key point is that we can disable this credential on cloud shell by setting this `isAvailable` to false. I can think of other less clean approaches, like throwing on getToken or throwing on some other place if this is the only MSI *available* and the user specified a client ID. The alternatives seem much less clean to me than just making this MSI unavailable if `clientId` is provided.

> The warning says "managed identities aren't available" in the cloud shell

Yeah this doesn’t look good! thank you for pointing it out. I have suggested some alternatives in another thread.

> appeared to work previously. Feels a bit like a breaking change

This is fundamentally a bug fix, because in the past providing a client ID in cloud shell would have ignored the client ID and instead picked an entirely different identity (the system identity). We don’t want to default to the system identity if the user provided a client ID.

Reply:
Well, I disagree that it's a bug fix. It's not making something broken work correctly; it's disabling a buggy thing by making it an error case. I think that's an important difference. In this case, someone could have functional code in which the client_id parameter just doesn't do anything. Or: they could be _testing_ something in Cloud Shell, but the code provides client_id for a prod use case that is deployed to App Service. Both of those cases now fail because no MSI provider is considered to be "available." I don't think we have good options here, but this also doesn't feel good.

On logging, I really feel that an info-level log is more appropriate, and maybe a warning-level log in `getToken` for this MSI provider if it is actually selected and `client_id` is set (I understand that in the current implementation in this PR, that couldn't happen, but if it were possible then a warning-level log would make sense). Consider someone who is using App Service MSI and passing client_id. Maybe that person forwards logs to Azure Monitor, so they have `AZURE_LOG_LEVEL=warning`. They will get warnings even though they never actually used cloud shell MSI if it was even _tested_ for availability. It feels a bit strange for the code to _warn_ the developer about that, but an info-level log ("hey FYI this parameter doesn't work with this provider") would be perfectly normal.

Reply:
> It's not making something broken work correctly

It is making something broken work correctly since if users passed the `clientId` on the cloud shell environment they would have authenticated with the wrong client.

Reply:
> someone could have functional code in which the client_id parameter just doesn't do anything

The fact that it seemed to work doesn’t mean that it really worked. It didn’t work since the client ID wasn’t being used.

> Or: they could be testing something in Cloud Shell, but the code provides client_id for a prod use case that is deployed to App Service

In this case their test was wrong, but the deployed service would continue to work.

Reply:
> Consider someone who is using App Service MSI and passing client_id. Maybe that person forwards logs to Azure Monitor, so they have AZURE_LOG_LEVEL=warning. They will get warnings even though they never actually used cloud shell MSI if it was even tested for availability. It feels a bit strange for the code to warn the developer about that, but an info-level log ("hey FYI this parameter doesn't work with this provider") would be perfectly normal.

Ok I like this reasoning! @KarishmaGhiya let’s make this an info log then.

Reply:
@KarishmaGhiya Wait, let’s bring this conversation with other languages. They’re using a warning, but perhaps not in the same place? Let’s make sure we pick info or warning for the right reason.

Reply:
I have some good news that will go well with your feedback, @witemple-msft !

After reading a conversation we had with the Identity crew, and this post by @KarishmaGhiya here: https://github.com/Azure/azure-sdk-for-js/issues/20142#issuecomment-1030501760 it is clear to me that this PR should be about adding a warning on getToken and not about preventing getToken 🙂 which means no arguably breaking changes.